gc/src/gc.rs

```diff
@@ -199,13 +214,23 @@ fn collect_garbage(st: &mut GcState) {
     unsafe fn mark(head: &Cell<Option<NonNull<GcBox<dyn Trace>>>>) -> Vec<Unmarked<'_>> {
         // Walk the tree, tracing and marking the nodes
         let mut mark_head = head.get();
-        while let Some(node) = mark_head {
-            if (*node.as_ptr()).header.roots() > 0 {
-                (*node.as_ptr()).trace_inner();
+        MARK_QUEUE.with(|queue| {
+            queue.borrow_mut().clear();
+            while let Some(node) = mark_head {
+                if (*node.as_ptr()).header.roots() > 0 {
+                    (*node.as_ptr()).trace_inner();
+                }
+
+                mark_head = (*node.as_ptr()).header.next.get();
             }
 
-            mark_head = (*node.as_ptr()).header.next.get();
-        }
+            while let Some(node) = {
+                let value = queue.borrow_mut().pop();
+                value
```

Comment:
This is just boilerplate to convince the compiler to drop the mutable borrow before calling into the `while` body (I don't know if there's a nicer/more idiomatic way to accomplish that)